File: pulse-java/src/main/java/com/fluxchess/pulse/java/model/Castling.java
/*
 * Copyright 2013-2023 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */
package com.fluxchess.pulse.java.model;

public final class Castling {

	public static final int WHITE_KINGSIDE = 1; // 1 << 0
	public static final int WHITE_QUEENSIDE = 1 << 1;
	public static final int BLACK_KINGSIDE = 1 << 2;
	public static final int BLACK_QUEENSIDE = 1 << 3;

	public static final int NOCASTLING = 0;

	public static final int VALUES_LENGTH = 16;

	private Castling() {
	}

	public static int valueOf(int color, int castlingtype) {
		switch (color) {
			case Color.WHITE:
				switch (castlingtype) {
					case CastlingType.KINGSIDE:
						return WHITE_KINGSIDE;
					case CastlingType.QUEENSIDE:
						return WHITE_QUEENSIDE;
					default:
						throw new IllegalArgumentException();
				}
			case Color.BLACK:
				switch (castlingtype) {
					case CastlingType.KINGSIDE:
						return BLACK_KINGSIDE;
					case CastlingType.QUEENSIDE:
						return BLACK_QUEENSIDE;
					default:
						throw new IllegalArgumentException();
				}
			default:
				throw new IllegalArgumentException();
		}
	}
}

File: pulse-go/internal/pulse/engine/castling.go
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package engine

import "fmt"

type Castling = int

const (
	WhiteKingside  Castling = 1 // 1 << 0
	WhiteQueenside Castling = 1 << 1
	BlackKingside  Castling = 1 << 2
	BlackQueenside Castling = 1 << 3

	NoCastling Castling = 0
)

var (
	Castlings = [4]Castling{
		WhiteKingside, WhiteQueenside,
		BlackKingside, BlackQueenside,
	}
)

func IsValidCastling(cast Castling) bool {
	switch cast {
	case WhiteKingside, WhiteQueenside, BlackKingside, BlackQueenside:
		return true
	default:
		return false
	}
}

func CastlingOf(col Color, ct CastlingType) Castling {
	switch col {
	case White:
		switch ct {
		case Kingside:
			return WhiteKingside
		case Queenside:
			return WhiteQueenside
		default:
			panic(fmt.Sprintf("Invalid castling type: %v", ct))
		}
	case Black:
		switch ct {
		case Kingside:
			return BlackKingside
		case Queenside:
			return BlackQueenside
		default:
			panic(fmt.Sprintf("Invalid castling type: %v", ct))
		}
	default:
		panic(fmt.Sprintf("Invalid color: %v", col))
	}
}

func CastlingColorOf(cast Castling) Color {
	switch cast {
	case WhiteKingside, WhiteQueenside:
		return White
	case BlackKingside, BlackQueenside:
		return Black
	default:
		panic(fmt.Sprintf("Invalid castling: %v", cast))
	}
}

func CastlingTypeOf(cast Castling) CastlingType {
	switch cast {
	case WhiteKingside, BlackKingside:
		return Kingside
	case WhiteQueenside, BlackQueenside:
		return Queenside
	default:
		panic(fmt.Sprintf("Invalid castling: %v", cast))
	}
}
